src/transformers/dynamic_module_utils.py

```diff
@@ -196,9 +197,15 @@ def get_class_in_module(class_name: str, module_path: Union[str, os.PathLike]) -
     Returns:
         `typing.Type`: The class looked for.
     """
-    name = os.path.normpath(module_path).replace(".py", "").replace(os.path.sep, ".")
-    module_path = str(Path(HF_MODULES_CACHE) / module_path)
-    module = importlib.machinery.SourceFileLoader(name, module_path).load_module()
+    name = os.path.normpath(module_path).rstrip(".py").replace(os.path.sep, ".")
+    module_spec = importlib.util.spec_from_file_location(name, location=Path(HF_MODULES_CACHE) / module_path)
+    module = sys.modules.get(name)
+    if module is None:
+        module = importlib.util.module_from_spec(module_spec)
+        # insert it into sys.modules before any loading begins
+        sys.modules[name] = module
+    # reload in both cases
+    module_spec.loader.exec_module(module)
```

Comment:
@XuehaiPan why do you call `exec_module` in both cases here?

it has the effect of reloading code definitions from disk for every model instantiation, which breaks monkey-patches applied onto trust_remote_code models (i.e. liger-kernel, etc)

Reply:
Hmm, I think here we did this to match the previous code block, where

> module = importlib.machinery.SourceFileLoader(name, module_path).load_module()

was always executed.

Reply:
> @XuehaiPan why do you call exec_module in both cases here?

This intentional. The module instance will be inserted into `sys.module` during initialization, whether it is executed or not (partially initializaed). Maybe we can add a `module.__initialized` indicator to ensure the module file is executed only once. I will submit a PR to resolve this.

Reply:
Addressed in #33162

- #33162